### ISSUE TITLE
Special attributes

### ISSUE THREAD

**Issue by erikgrinaker** [sanity-io/gradient#601]
We already have the `^` "operator", referring to the parent document. We will also need e.g. `@` for "this" (the current document). These are not really operators, since they do not operate on anything. I think we should formalize and generalize this concept.

I suggest naming these special attributes (or predefined/magic/whatever), with the same semantics as regular attributes. This implies that they are document-scoped (available in e.g. filters), and that they can be chained - as a contrived example, `*[ @.@.value == ^.@.^.value ].^.^.@.value` should be valid and meaningful.

We should take care to name these attributes using symbols that are not valid in regular attributes, to avoid name clashes.

**Comment by erikgrinaker** [comment]
I suppose * would also be a special attribute, although globally scoped. Not sure how it fits in here, but I feel like it should be part of the same family of objects.

**Comment by atombender** [comment]
Disagree about the above. If `@` means `this`, then `@.@` makes no sense. In other languages such as JavaScript `this.this` means the member `this` on `this`:

```
> this.this = 1;
1
> this.this
1
```

`^` can arguably only occur first in an expression because it doesn't refer to objects, but the (lexical) scope of the query. You can't do `foo.^`.

Duplicate parents intermingled with attributes have to resolve to `^`. For example `^.foo.^` just means `^` and can be collapsed to it. It should arguably not be permitted.

**Comment by erikgrinaker** [comment]
True, it would only make sense for e.g. other->^, which is ugly and less useful. I still think we need an organizing concept for these objects, but I agree that attributes is the wrong one.

**Comment by atombender** [comment]
`something->^` doesn't make sense to me.
